signer/jws.go

```diff
@@ -0,0 +1,475 @@
+	if err != nil {
+		return nil, err
+	}
+
+	b, err := json.Marshal(j)
+	if err != nil {
+		return nil, err
+	}
+	jws.internalEnv = j
+	return b, nil
+}
+
+func (jws *JWSEnvelope) getSignerInfo() (SignerInfo, error) {
+	signInfo := SignerInfo{}
+	if reflect.DeepEqual(jws.internalEnv, newJwsInternalEnvelope()) {
+		return signInfo, SignatureNotFoundError{}
+	}
+
+	// parse payload
+	if payload, err := base64.RawURLEncoding.DecodeString(jws.internalEnv.Payload); err != nil {
+		return signInfo, err
+	} else {
+		signInfo.Payload = payload
+	}
+
+	// parse protected headers
+	if protected, err := base64.RawURLEncoding.DecodeString(jws.internalEnv.Protected); err != nil {
+		return signInfo, err
+	} else {
+		var pHeaders map[string]interface{}
+		if err = json.Unmarshal(protected, &pHeaders); err != nil {
+			return signInfo, err
+		}
+		if err := populateProtectedHeaders(pHeaders, &signInfo); err != nil {
+			return signInfo, err
+		}
+	}
+
+	// parse signature
+	if sig, err := base64.RawURLEncoding.DecodeString(jws.internalEnv.Signature); err != nil {
+		return signInfo, err
+	} else {
+		signInfo.Signature = sig
+	}
+
+	// parse headers
+	certs := make([]x509.Certificate, 0)
+	for _, certBytes := range jws.internalEnv.Header.CertChain {
+		if cert, err := x509.ParseCertificate(certBytes); err != nil {
+			return signInfo, err
+		} else {
+			certs = append(certs, *cert)
+		}
+	}
+	signInfo.CertificateChain = certs
+	signInfo.UnsignedAttributes.SigningAgent = jws.internalEnv.Header.SigningAgent
+	signInfo.TimestampSignature = jws.internalEnv.Header.TimestampSignature
+
+	return signInfo, nil
+}
+
+func populateProtectedHeaders(pHeaders map[string]interface{}, signInfo *SignerInfo) error {
+	crit, err := getAndValidateCriticalHeaders(pHeaders)
+	if err != nil {
+		return err
+	}
+	if err := populateAlg(pHeaders, &signInfo.SignatureAlgorithm); err != nil {
+		return err
+	}
+	if err := populateString(pHeaders, ctyHeaderKey, &signInfo.PayloadContentType); err != nil {
+		return err
+	}
+	if err := populateTime(pHeaders, signingTimeHeaderKey, &signInfo.SignedAttributes.SigningTime); err != nil {
+		return err
+	}
+	if err := populateTime(pHeaders, expiryHeaderKey, &signInfo.SignedAttributes.Expiry); err != nil {
+		return err
+	}
+
+	// This should be last entry and populates crit and other protected headers
+	if err := populateExtendedAttributes(pHeaders, crit, &signInfo.SignedAttributes.ExtendedAttributes); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func populateString(data map[string]interface{}, s string, holder *string) error {
+	if val, ok := data[s]; ok {
+		valString, ok := val.(string)
+		if !ok {
+			return MalformedSignatureError{msg: fmt.Sprintf("failed to parse %q in protected headers", s)}
+		}
+		*holder = valString
+		delete(data, s)
+	}
+	return nil
+}
+
+func populateAlg(data map[string]interface{}, holder *signatureAlgorithm) error {
+	var alg string
+	if err := populateString(data, "alg", &alg); err != nil {
+		return err
```

Comment:
Do we need to wrap these errors into `MalformedSignatureError`?

Reply:
Not required as its taken care by https://github.com/priteshbandi/notation-core-go/blob/core-refactor/signer/signer.go#L139